### PR TITLE
feat: Publish progress of processing the bagfile

### PR DESCRIPTION
**Pull request by alireza-hosseini**
## What I did
- Added a publisher which is publishing the progress of processing the bag-files
  - The message type is `cartographer_ros_msgs/BagfileProgress` which is added in this PR
    - It contains number of bagfiles in the queue, current bagfile name and Id, total and remaining duration of procssing the bagfile (in seconds) and also total and remaining number of processed messages  
 - The time interval of publishing the progress can be set using **~bagfile_progress_pub_interval** ROS param. The default value is **10.0** seconds.
- This is mostly useful for cases that the Cartographer progress needs to be monitored or visualized. 
 
## How I tested
In simulation

### CODE REVIEW

**Comment by alireza-hosseini** [comment]
I signed it!

File: cartographer_ros/cartographer_ros/playable_bag.cc [hunk 1]
```diff
@@ -62,10 +65,12 @@ rosbag::MessageInstance PlayableBag::GetNextMessage() {
   const rosbag::MessageInstance msg = buffered_messages_.front();
   buffered_messages_.pop_front();
   AdvanceUntilMessageAvailable();
+  float processed_seconds = (msg.getTime() - view_->getBeginTime()).toSec();
+  float progress = processed_seconds / duration_in_seconds_;
```
Comment (MichaelGrupp) [inline cartographer_ros/cartographer_ros/playable_bag.cc]: double?

File: cartographer_ros/cartographer_ros/playable_bag.cc [hunk 2]
```diff
@@ -62,10 +65,12 @@ rosbag::MessageInstance PlayableBag::GetNextMessage() {
...
+  bag_progress_pub_.publish(progress);
```
[the remark this answers is not in the document]
Comment (alireza-hosseini) [reply cartographer_ros/cartographer_ros/playable_bag.cc]: @MichaelGrupp I added a ROS param for setting the time interval of publishing the progress.

File: cartographer_ros/cartographer_ros/playable_bag.h [hunk 3]
```diff
@@ -49,6 +50,10 @@ class PlayableBag {
   void AdvanceOneMessage();
   void AdvanceUntilMessageAvailable();
 
+  ros::NodeHandle pnh_;
+  //!@brief Publishes the progress of processing the current bag file [0.0, 1.0]
+  ros::Publisher bag_progress_pub_;
```
[the remark this answers is not in the document]
Comment (alireza-hosseini) [reply cartographer_ros/cartographer_ros/playable_bag.h]: Fixed.

File: cartographer_ros/cartographer_ros/playable_bag.cc (hunk 1 above)
Comment (alireza-hosseini) [reply cartographer_ros/cartographer_ros/playable_bag.cc]: Fixed.

**Comment by alireza-hosseini** [comment]
@MichaelGrupp Thanks for your review, I am totaly agree with you, I added a new ROS message containing more detailed information about the progress.
Please read the updated PR explanation for more details and also review new changes.

**Comment by MichaelGrupp** [comment]
Ok! I'm just a contributor and not a maintainer, so I'll leave the review to them now ;)

File: cartographer_ros/cartographer_ros/playable_bag.cc [hunk 4]
```diff
@@ -57,16 +57,24 @@ std::tuple<ros::Time, ros::Time> PlayableBag::GetBeginEndTime() const {
...
   CHECK(IsMessageAvailable());
   const rosbag::MessageInstance msg = buffered_messages_.front();
   buffered_messages_.pop_front();
   AdvanceUntilMessageAvailable();
+  double processed_seconds = (msg.getTime() - view_->getBeginTime()).toSec();
   if ((log_counter_++ % 10000) == 0) {
-    LOG(INFO) << "Processed " << (msg.getTime() - view_->getBeginTime()).toSec()
-              << " of " << duration_in_seconds_ << " seconds of bag "
-              << bag_filename_;
+    LOG(INFO) << "Processed " << processed_seconds << " of "
```
[the remark this answers is not in the document]
Comment (alireza-hosseini) [reply cartographer_ros/cartographer_ros/playable_bag.cc]: Nice feature, Please take a look at the new commit.

[the remark this answers is not in the document]
Comment (ojura) [reply cartographer_ros/cartographer_ros/playable_bag.cc]: `LOG_EVERY_N` was intentionally not used there. If you have multiple concurrent bags, `LOG_EVERY_N` would not nicely periodically output progress from each bag, but rather cause different `PlayableBags` to use the same log counter.

[the remark this answers is not in the document]
Comment (alireza-hosseini) [reply cartographer_ros/cartographer_ros/playable_bag.cc]: @ojura I think your right (however I couldn't really simulate this case on my laptop), so I'm going to drop the commit.

**Comment by alireza-hosseini** [comment]
@ojura I couldn't really simulate the concurrent bagfile processing case on my machine and from what I understood, bagfiles are being processed sequentially. Am I missing something?

**Comment by alireza-hosseini** [comment]
@cschuet @wohe Since I created this PR nearly one month ago, I'm wondering if you have any plan to review this PR or not, due to the fact that it is kind of blocking me to progress on my current project.

**Comment by ojura** [comment]
@alireza-hosseini in `--bag-files` you specify two (or more) bag files from two robots which were driving around at the same time. Look at #636 for more details.

File: cartographer_ros/cartographer_ros/playable_bag.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (alireza-hosseini) [reply cartographer_ros/cartographer_ros/playable_bag.cc]: @ojura I dropped this commit.

**Comment by alireza-hosseini** [comment]
Using the example bagfiles you provided on #636 I could finally test the simultaneous processing mode and you were totally right. I added a `std::map` to store last progress time of each bagfile and I could see the progress of **alpha** and **delta** bagfiles concurrently with an accurate time interval.

**Comment by ojura** [comment]
Great! Collating concurrent bags is the whole point of the`PlayableBagMultiplexer` class.  One small suggestion: why not store the time inside each `PlayableBag`, just like `log_counter_`, so you don‘t have to keep a map? And do publishing from `PlayableBag::GetNextMessage‘?

**Comment by alireza-hosseini** [comment]
@ojura Thanks, actually I was publishing the progress from `PlayableBag` at the very beginning of this PR (you can see the discussions above), but there we don't have access to the total number of bag files which is a useful data. That's why I moved the publishing part to `PlayableBagMultiplexer` class.

**Comment by wohe** [comment]
Sorry, this is the first time I see this PR, but I currently also do not have the time to review this. I'll leave it to @cschuet.

**Comment by alireza-hosseini** [comment]
@pifon2a Are you ok with the changes?

File: cartographer_ros/cartographer_ros/playable_bag.cc [hunk 5]
```diff
@@ -57,16 +57,24 @@ std::tuple<ros::Time, ros::Time> PlayableBag::GetBeginEndTime() const {
   return std::make_tuple(view_->getBeginTime(), view_->getEndTime());
 }
 
-rosbag::MessageInstance PlayableBag::GetNextMessage() {
+rosbag::MessageInstance PlayableBag::GetNextMessage(
+    cartographer_ros_msgs::BagfileProgress& progress) {
```
Comment (pifon2a) [inline cartographer_ros/cartographer_ros/playable_bag.cc]: `(cartographer_ros_msgs::BagfileProgress* progress)`. Pass mutable parameters by pointer.

Comment (alireza-hosseini) [reply cartographer_ros/cartographer_ros/playable_bag.cc]: done.

**Comment by pifon2a** [comment]
@wally-the-cartographer merge

**Comment by wally-the-cartographer** [comment]
Merge requested by authorized user pifon2a. Merge queue now has a length of 1.

**Comment by wally-the-cartographer** [comment]
Refusing to merge, since the combined checks state is failure

**Comment by alireza-hosseini** [comment]
@pifon2a CI build failed with following error:

> The job exceeded the maximum time limit for jobs, and has been terminated.
Is there any way that I can retrigger the build?

**Comment by pifon2a** [comment]
@alireza-hosseini Retriggered the build. I'll make sure it gets submitted today. No worries.

**Comment by alireza-hosseini** [comment]
@pifon2a Is there any update regarding the melodic build issue?

**Comment by ojura** [comment]
@alireza-hosseini try `git commit -m"Re-run Travis." --allow-empty && git push`

**Comment by alireza-hosseini** [comment]
@pifon2a Can you please merge this PR?

**Comment by pifon2a** [comment]
@wally-the-cartographer merge

**Comment by alireza-hosseini** [comment]
@pifon2a It seems something went wrong during merge.

**Comment by pifon2a** [comment]
@alireza-hosseini I think Wally is dead.

**Comment by alireza-hosseini** [comment]
@pifon2a I don't know who is the maintainer of the @wally-the-cartographer so I could communicate the problem to him/her. I guess @cschuet ?

**Comment by SirVer** [comment]
@wally-the-cartographer merge

**Comment by alireza-hosseini** [comment]
@pifon2a Thanks a lot for taking the time and effort to merge this PR.

**Comment by ojura** [comment]
I would like to report that this PR completely trashed the performance of the offline node for me. The most alarming is that with `-skip_seconds`, which previously used just under half a second to get through the first e.g. 1000 seconds of a bag, now takes 10 minutes or more (and this is without calling `HandleMessage`, since all messages are being skipped!). And it keeps slowing down more and more.

@alireza-hosseini can you please revert this PR and investigate?

**Comment by alireza-hosseini** [comment]
@ojura Since you have a proper test case, can you please try removing the line https://github.com/googlecartographer/cartographer_ros/blob/829e2dc43f087668c0ae0cf8613c21ef83acf7fe/cartographer_ros/cartographer_ros/playable_bag.cc#L74 and try again? It is the fastest way to fix the problem. Thanks.

**Comment by ojura** [comment]
@alireza-hosseini no special test case needed, just take any dataset and set `-skip_seconds` to something larger than the bag length. The result should be offline node "processing" the bag very quickly.

**Comment by ojura** [comment]
Yes, exactly! So the total complexity of stepping through the bag is O(n^2), which really kills it.

Doesn't `log_counter_` that we previously discussed here contain the exact number you are looking for? It might be a good idea to rename it to `message_counter_` so it's more clear.

**Comment by alireza-hosseini** [comment]
@ojura I created the PR (#1045) containing the fixes, would be great if you take a look.

**Pull request merged**